Sources/PT.PM.Common/MessagePack/RootUstMessagePackSerializer.cs

```diff
@@ -69,52 +73,423 @@ public static byte[] Serialize(RootUst rootUst, bool isLineColumn, bool compress
+            }
 
             if (type == typeof(TextSpan))
             {
-                return (IMessagePackFormatter<T>) textSpanFormatter;
+                return textSpanSerializer.Serialize(ref bytes, offset, (TextSpan) value);
             }
 
-            if (type == typeof(TextFile))
+            if (type == typeof(double))
             {
-                return (IMessagePackFormatter<T>) fileFormatter;
+                return MessagePackBinary.WriteDouble(ref bytes, offset, (double) value);
             }
 
-            if (type == typeof(RootUst))
+            if (type.IsEnum)
             {
-                return (IMessagePackFormatter<T>) rootUstFormatter;
+                return MessagePackBinary.WriteInt32(ref bytes, offset, (int) value);
             }
 
-            return StandardResolver.Instance.GetFormatter<T>();
+            Type[] typeInterfaces = type.GetInterfaces();
+
+            if (typeInterfaces.Contains(typeof(IFile)))
+            {
+                return fileSerializer.Serialize(ref bytes, offset, (IFile) value);
+            }
+
+            if (typeInterfaces.Contains(typeof(IList)))
+            {
+                int newOffset = offset;
+                IList collection = (IList) value;
+                newOffset += MessagePackBinary.WriteArrayHeader(ref bytes, newOffset, collection.Count);
+                foreach (object item in collection)
+                {
+                    newOffset += SerializeObject(ref bytes, newOffset, item?.GetType(), item);
+                }
+
+                return newOffset - offset;
+            }
+
+            if (type == typeof(char))
+            {
+                return MessagePackBinary.WriteChar(ref bytes, offset, (char) value);
+            }
+
+            if (type == typeof(byte))
+            {
+                return MessagePackBinary.WriteByte(ref bytes, offset, (byte) value);
+            }
+
+            if (type == typeof(sbyte))
+            {
+                return MessagePackBinary.WriteSByte(ref bytes, offset, (sbyte) value);
+            }
+
+            if (type == typeof(short))
+            {
+                return MessagePackBinary.WriteInt16(ref bytes, offset, (short) value);
+            }
+
+            if (type == typeof(ushort))
+            {
+                return MessagePackBinary.WriteUInt16(ref bytes, offset, (ushort) value);
+            }
+
+            if (type == typeof(ulong))
+            {
+                return MessagePackBinary.WriteUInt64(ref bytes, offset, (ulong) value);
+            }
+
+            if (type == typeof(float))
+            {
+                return MessagePackBinary.WriteSingle(ref bytes, offset, (float) value);
+            }
+
+            if (type == typeof(DateTime))
+            {
+                return MessagePackBinary.WriteDateTime(ref bytes, offset, (DateTime) value);
+            }
+
+            throw new NotImplementedException($"Serialization of {type.Name} is not supported");
```

Comment:
No need for `uint`s? :smiley:

Reply:
Not now, but later it's possible. I did it just in case.

Reply:
Sorry, I've read it inattentively. You are right: `uint` is missing.